### ISSUE TITLE
Attaching SSL thumbprint to existing CustomHostnameBinding breaks state, refresh does not help

### ISSUE THREAD

**Issue by scp-mb** [pulumi/pulumi-azure#751]
It seems there may be an issue with replacing custom hostname bindings. The process I followed was as below:

Using Pulumi v2.12.1 with Azure provider 3.28.0

1) Create new custom hostname binding without SSL thumbprint
2) Pulumi up
3) Update code to specify thumbprint
4) Pulumi up, at this point the replacement fails as the resource already exists (no unique name on hostname bindings)
5) Applied DeleteBeforeReplace = true
6) Pulumi up, hostname binding gets deleted successfully and replacement created according to the CLI.

At this point everything should work, however checking in Azure the hostname binding is completely missing, with Pulumi's state thinking it exists. Running `pulumi refresh` doesn't pick up any changes even though the binding is completely gone from Azure.

**Comment by scp-mb** [comment]
It seems I've managed to fix up my state by removing the resource in question from the state with `pulumi state rm`, adding the resource in Azure (without SSL), reimporting it and then making the SSL thumbprint modification, this time with no issues.

I'm not sure why it didn't work the first time, and the fact that `pulumi refresh` didn't pick up the missing resource is still concerning.

If helpful, here is a snippet from the state of the resource in question with information redacted where necessary. I notice there is a `pendingReplacement` field set to true, is this the cause of the issue? If so, no logs or errors were raised to suggest so.

```
{
    "urn": "urn:pulumi:uat::Company::azure:appservice/customHostnameBinding:CustomHostnameBinding::comments.domain.net-secondary",
    "custom": true,
    "id": "/subscriptions/redacted/resourceGroups/redacted-resource-group/providers/Microsoft.Web/sites/redacted-app-comments-api-secondary/hostNameBindings/comments.domain.net",
    "type": "azure:appservice/customHostnameBinding:CustomHostnameBinding",
    "inputs": {
        "__defaults": [],
        "appServiceName": "redacted-app-comments-api-secondary",
        "hostname": "comments.domain.net",
        "resourceGroupName": "redacted-resource-group"
    },
    "outputs": {
        "__meta": "{\"e2bfb730-ecaa-11e6-8f88-34363bc7c4c0\":{\"create\":1800000000000,\"delete\":1800000000000,\"read\":300000000000,\"update\":1800000000000}}",
        "appServiceName": "redacted-app-comments-api-secondary",
        "hostname": "comments.domain.net",
        "id": "/subscriptions/redacted/resourceGroups/redacted-resource-group/providers/Microsoft.Web/sites/redacted-app-comments-api-secondary/hostNameBindings/comments.domain.net",
        "resourceGroupName": "redacted-resource-group",
        "sslState": "",
        "thumbprint": "",
        "virtualIp": ""
    },
    "parent": "urn:pulumi:uat::Company::pulumi:pulumi:Stack::Company-uat",
    "dependencies": [
        "urn:pulumi:uat::Company::azure:appservice/appService:AppService::redacted-app-comments-api-secondary",
        "urn:pulumi:uat::Company::azure:dns/txtRecord:TxtRecord::comments-asuid"
    ],
    "provider": "urn:pulumi:uat::Company::pulumi:providers:azure::default_3_28_0::cda4cfc1-22ae-45d2-8cb9-347c4416eb65",
    "propertyDependencies": {
        "appServiceName": null,
        "hostname": null,
        "resourceGroupName": null
    },
    "pendingReplacement": true,
    "importID": "/subscriptions/redacted/resourceGroups/redacted-resource-group/providers/Microsoft.Web/sites/redacted-app-comments-api-secondary/hostNameBindings/comments.domain.net"
},
```

**Comment by scp-mb** [comment]
As a bit of a follow up to this, we're currently updating the SSL certificate thumbprint on ~100 hostname bindings and it is **painfully** slow, it almost seems like with `DeleteBeforeReplace` enabled it seems it will delete the hostnames as soon as it can without waiting for the previous replacement to be created, but then creation only happens sequentially, causing a not insignificant amount of downtime the further down the list of hostnames you get.

![](https://i.imgur.com/0XKXkRk.gif)

**Comment by scp-mb** [comment]
Well, after batching up the changes to only a few per `pulumi up` run to avoid excessive downtime, the same thing has happened again. It's deleted the hostname bindings, failed to create the replacements and the state is now broken where it thinks they exist and a `pulumi refresh` won't fix it. 

Leaving infrastructure in a half broken state and requiring hefty manual intervention is **really** bad. 😟 

I've captured the output from `pulumi up` this time around:

```

     Type                                       Name                               Status                   Info
     pulumi:pulumi:Stack                        <redacted>-prod-tenanted            **failed**               1 error
 +-  ├─ azure:appservice:CustomHostnameBinding  <redacted>                          **replacing failed**     [diff: ~thumbprint]; 1 error
 +-  ├─ azure:appservice:CustomHostnameBinding  <redacted>                          **replacing failed**     [diff: ~thumbprint]; 1 error
 +-  ├─ azure:appservice:CustomHostnameBinding  <redacted>                          **replacing failed**     [diff: ~thumbprint]; 1 error
 +-  ├─ azure:appservice:CustomHostnameBinding  <redacted>                          **replacing failed**     [diff: ~thumbprint]; 1 error
 +-  ├─ azure:appservice:CustomHostnameBinding  <redacted>                          **replacing failed**     [diff: ~thumbprint]; 1 error
 +-  └─ azure:appservice:CustomHostnameBinding  <redacted>                          **replacing failed**     1 error

Diagnostics:
  azure:appservice:CustomHostnameBinding (<redacted>):
    error: deleting urn:pulumi:prod-tenanted::<redacted>::azure:appservice/customHostnameBinding:CustomHostnameBinding::<redacted>: Error deleting Custom Hostname Binding "<redacted>" (App Service "<redacted>" / Resource Group "<redacted>"): web.AppsClient#DeleteHostNameBinding: Failure sending request: StatusCode=409 -- Original Error: autorest/azure: Service returned an error. Status=<nil> <nil>

  azure:appservice:CustomHostnameBinding (<redacted>):
    error: Error creating/updating Custom Hostname Binding "<redacted>" (App Service "<redacted>" / Resource Group "<redacted>"): web.AppsClient#CreateOrUpdateHostNameBinding: Failure sending request: StatusCode=409 -- Original Error: autorest/azure: Service returned an error. Status=<nil> <nil>

  azure:appservice:CustomHostnameBinding (<redacted>):
    error: Error creating/updating Custom Hostname Binding "<redacted>" (App Service "<redacted>" / Resource Group "<redacted>"): web.AppsClient#CreateOrUpdateHostNameBinding: Failure sending request: StatusCode=409 -- Original Error: autorest/azure: Service returned an error. Status=<nil> <nil>

  azure:appservice:CustomHostnameBinding (<redacted>):
    error: Error creating/updating Custom Hostname Binding "<redacted>" (App Service "<redacted>" / Resource Group "<redacted>"): web.AppsClient#CreateOrUpdateHostNameBinding: Failure sending request: StatusCode=409 -- Original Error: autorest/azure: Service returned an error. Status=<nil> <nil>

  pulumi:pulumi:Stack (<redacted>-prod-tenanted):
    error: update failed

  azure:appservice:CustomHostnameBinding (<redacted>):
    error: Error creating/updating Custom Hostname Binding "<redacted>" (App Service "<redacted>" / Resource Group "<redacted>"): web.AppsClient#CreateOrUpdateHostNameBinding: Failure sending request: StatusCode=409 -- Original Error: autorest/azure: Service returned an error. Status=<nil> <nil>

  azure:appservice:CustomHostnameBinding (<redacted>):
    error: Error creating/updating Custom Hostname Binding "<redacted>" (App Service "<redacted>" / Resource Group "<redacted>"): web.AppsClient#CreateOrUpdateHostNameBinding: Failure sending request: StatusCode=409 -- Original Error: autorest/azure: Service returned an error. Status=<nil> <nil>

Resources:
    +-5 replaced
    411 unchanged
```

**Comment by scp-mb** [comment]
It's come to that time where I've had to to a bulk update on hostname bindings again so I figured I'd try the nextgen provider this time and see if it's any better.

Spoiler: It isn't. Two resources got deleted and replaced, the all the others failed with a 409 conflict in the Azure portal logs because a pending operation is already in progress. It's also managed to get it into a state where Azure thinks an operation is still ongoing and I can't see the resources in the portal that it did apparently replace.

Managing hostname bindings is essentially completely broken if you have to do anything around SSL certificates in both providers.

@komalali Do you know if anyone will be taking a look at this, otherwise I'm going to have to scrap the idea of using Pulumi to manage hostname bindings entirely.

**Comment by viveklak** [comment]
@scp-mb could you share some of the code? A separate ticket for the nextgen with a basic repro would really help to investigate this!

**Comment by scp-mb** [comment]
@viveklak It's a bit awkward to make a shareable POC due to requiring some resources specific to each environment such a DNS zones and multiple SSL certificates, but I can outline the general reproduction steps

1) Create a Resource group
2) Create app service plan that supports custom hostnames
3) Create app service
4) Create a DNS zone and some CNAME records pointing to the new app service
5) Add a decent (5-10+) amount of hostname bindings to the app service matching the previous CNAME records, with a SSL certificate attached to the binding using `SslState = SniEnabled

Once all that is in place, imagine you're coming up to renewal time for your certificate, or you need to revoke & replace it

As such, you change the `Thumbprint` property on your `CustomHostnameBidningArgs`. At this point running `pulumi up` will result in the above issues.

The Azure portal will report the error as a 409 conflict as an existing operation is already in progress
![image](https://user-images.githubusercontent.com/52526733/103802909-c43a7e80-5047-11eb-8333-576a3599b9f8.png)

A redacted snippet of the `CustomHostnameBinding` is below:

```
new Pulumi.Azure.AppService.CustomHostnameBinding(<pulumi-resource-id>,
                new CustomHostnameBindingArgs
                {
                    ResourceGroupName = <resource-group-name>,
                    AppServiceName = <app-service-name>,
                    Hostname = <hostname>,
                    SslState = "SniEnabled",
                    Thumbprint = <certificate-thumbprint>
                },
                new CustomResourceOptions
                {
                    DeleteBeforeReplace = true
                });
```

As a workaround, I've discovered you can bulk update hostname bindings via the following endpoint which is what the Azure portal uses when updating a SSL certificate thumbprint on a hostname binding, and with some modification of the request allows you to update multiple by setting the `toUpdate` flag on each object

Interestingly, this seems to be an in-place update with no downtime which both providers don't seem to currently do.

https://docs.microsoft.com/en-us/rest/api/appservice/webapps/update

**Comment by viveklak** [comment]
Thanks @scp-mb - will try to repro.

**Comment by scp-mb** [comment]
@viveklak Did you manage to have any luck with a repro?
